### ISSUE TITLE
Counterparty tries to spend an output already spent in unconfirmed tx

### ISSUE THREAD

**Issue by arhuaco** [CounterpartyXCP/counterparty-lib#811]
Hello there.

I have a few unit tests with Counterparty in regtest. Today one of them failed and I could "fix" it by mining a block before doing a TX. I checked the transactions and I think counterparty-lib is trying to spend an output that has already been spent in the mempool (not confirmed yet).

This TX is in the mempool:

    {
       "lock_time":0,
       "size":264,
       "inputs":[
          {
             "prev_out":{
                "index":1,
                "hash":"b4a0e795ee17c56fb4c3886dff87f9463a1b927c1147a3b2a7ec52c7f08b89cf"
             },
    "script":"47304402206b8ebfee6dba49c913f58dec7bbf52b17816d2d95bc7c7978c72f41c2fb8024602206019913c3bd273d9a2c0daa6de3a9d1810ee960ed4bf216d47b822eeb341fe360121039b67d69621634b64a7d7f4be5f0f95daae6d1984a56bfc4857e5a169bffc93de"
          }
       ],
       "version":1,
       "vin_sz":1,
       "hash":"342f973147a005d4d2fc8113824c1a7130b5637193375d93f03a79ad1a26130c",
       "vout_sz":3,
       "out":[
          {
             "script_string":"OP_DUP OP_HASH160 ef2034e33c1f9c5b7a4bc919f21f583398fe2492 OP_EQUALVERIFY OP_CHECKSIG",
             "address":"1NoP5bqskewdT7XoPs8ckJJtjU6HbSxivf",
             "value":5430,
             "script":"76a914ef2034e33c1f9c5b7a4bc919f21f583398fe249288ac"
          },
          {
             "script_string":"OP_RETURN b7cc839c5701ed91e007dccd9b53d45bc5e24e245782d0b18c61beb0",
             "value":0,
             "script":"6a1cb7cc839c5701ed91e007dccd9b53d45bc5e24e245782d0b18c61beb0"
          },
          {
             "script_string":"OP_DUP OP_HASH160 e2752565f2366653646da2ecd0cb278790d0e751 OP_EQUALVERIFY OP_CHECKSIG",
         "address":"1MeQ3on1EusqjPdNc7cqrsGHe4ymNX6h2S",
         "value":9764567,
         "script":"76a914e2752565f2366653646da2ecd0cb278790d0e75188ac"
          }
       ]
    }

And this one is created with create_send (BTC asset). I think the output used here is spent (or in limbo) in the previous unconfirmed transaction.

    {
       "lock_time":0,
       "size":226,
       "inputs":[
          {
             "prev_out":{
                "index":1,
                "hash":"b4a0e795ee17c56fb4c3886dff87f9463a1b927c1147a3b2a7ec52c7f08b89cf"
             },
             "script":"483045022100a2eeda715d4cb39d2304a06bb5778b8ca3125644eb86a6d48fba186d2e3fc01f022034e7e6e50f955bc94b12a24e162613b796950a11b71988d4003af966eac1167e0121039b67d69621634b64a7d7f4be5f0f95daae6d1984a56bfc4857e5a169bffc93de"
          }
       ],
       "version":1,
       "vin_sz":1,
       "hash":"f90b6723a8d5d9adbeaa4d8973609357517c229996bc0d132e8350b81ad577b0",
       "vout_sz":2,
       "out":[
          {
             "script_string":"OP_DUP OP_HASH160 ef2034e33c1f9c5b7a4bc919f21f583398fe2492 OP_EQUALVERIFY OP_CHECKSIG",
             "address":"1NoP5bqskewdT7XoPs8ckJJtjU6HbSxivf",
             "value":10000,
             "script":"76a914ef2034e33c1f9c5b7a4bc919f21f583398fe249288ac"
          },
          {
             "script_string":"OP_DUP OP_HASH160 e2752565f2366653646da2ecd0cb278790d0e751 OP_EQUALVERIFY OP_CHECKSIG",
             "address":"1MeQ3on1EusqjPdNc7cqrsGHe4ymNX6h2S",
             "value":9960000,
             "script":"76a914e2752565f2366653646da2ecd0cb278790d0e75188ac"
          }
       ]
    }

**Comment by rubensayshi** [comment]
related to https://github.com/CounterpartyXCP/counterwallet/issues/626#issuecomment-93486673

**Comment by rubensayshi** [comment]
it's because counterparty-lib only keeps track of 0conf transactions that are XCP transactions and ignores all normal BTC transactions.

**Comment by adamkrellenstein** [comment]
@robby-dermody, can you look at this?

**Comment by robby-dermody** [comment]
@adamkrellenstein : @rubensayshi is correct. We discussed this in the slack channel this morning, and I believe both think that adding in Cached mempool tracking for Non CP transactions should also be possible (I misinterpreted the design parameters around this when I did the initial optimizations, but we can tweak them for the next release)
 @rubensayshi I saw you were working on a solution? Should I hold tight until you finish your testing with that?
